# Lab 04 - pandas II

Name: Jake Schaeffer  
Class: CSCI 349 - Intro to Data Mining  
Semester: Spring 2020  
Instructor: Brian King

pandas documentation used as a reference for most(if not all) exercises

In [2]:
import sys
import numpy as np
import pandas as pd

In [17]:
days = ["Mon","Tue","Wed","Thurs","Fri","Student"]
scores = pd.DataFrame([pd.Series([9.5, 8.75, 8, 10, 7.75, "Bob"], index=days, name="week_1"),
pd.Series([9, 8, 10, 8.75, 7.25, "Jane"], index=days, name="week_1"),
pd.Series([8.5, 0, 9.75, 9, 6, "Bob"], index=days, name="week_2"),
pd.Series([7, 8.25, 9.25, 0, 8, "Jane"], index=days, name="week_2"),
pd.Series([7, 8.5, 9.25, 0, 0, "Bob"], index=days, name="week_3"),
pd.Series([8.5, 8.25, 9, 8, 7.5, "Jane"], index=days, name="week_3"),
pd.Series([9, 7, 8.5, 8, 8, "Bob"], index=days, name="week_4"),
pd.Series([9.25, 9.25, 8.5, 7.75, 7.5, "Jane"], index=days,name="week_4"),])
scores

Mon   Tue    Wed  Thurs   Fri Student
week_1  9.50  8.75   8.00  10.00  7.75     Bob
week_1  9.00  8.00  10.00   8.75  7.25    Jane
week_2  8.50  0.00   9.75   9.00  6.00     Bob
week_2  7.00  8.25   9.25   0.00  8.00    Jane
week_3  7.00  8.50   9.25   0.00  0.00     Bob
week_3  8.50  8.25   9.00   8.00  7.50    Jane
week_4  9.00  7.00   8.50   8.00  8.00     Bob
week_4  9.25  9.25   8.50   7.75  7.50    Jane

### 1) [P] Report the shape, the number of dimensions, the size of the data, and the data types of each column. Print each separately. Then create an additional code cell that shows the results of scores.info().

In [18]:
print("scores shape: ",scores.shape)
print("scores dimensions: ",scores.ndim)
print("scores size: ",scores.size)

# i feel like there is  a better way to do this in one line...
for i in range(len(days)):
    print("scores column {} type: {}".format(days[i],scores.iloc[:,i].dtype))

scores shape:  (8, 6)
scores dimensions:  2
scores size:  48
scores column Mon type: float64
scores column Tue type: float64
scores column Wed type: float64
scores column Thurs type: float64
scores column Fri type: float64
scores column Student type: object


In [19]:
scores.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8 entries, week_1 to week_4
Data columns (total 6 columns):
Mon        8 non-null float64
Tue        8 non-null float64
Wed        8 non-null float64
Thurs      8 non-null float64
Fri        8 non-null float64
Student    8 non-null object
dtypes: float64(5), object(1)
memory usage: 448.0+ bytes


### 2) [P] Show two different ways to report the number of observations in scores.

In [28]:
print("number of observations: ",len(scores.index))
print("number of observations: ",scores.shape[0])

number of observations:  8
number of observations:  8


### 3) [P] Show the variables (i.e. columns) in scores

In [9]:
scores.columns

Index(['Mon', 'Tue', 'Wed', 'Thurs', 'Fri', 'Student'], dtype='object')

### 4) [P] There are repetitive names in the index. This happens quite a bit, depending on how the rows are to be named. For now. print the unique index names in scores

In [10]:
scores.index.drop_duplicates()

Index(['week_1', 'week_2', 'week_3', 'week_4'], dtype='object')

### 5) [P] Rename the 'Thurs' column header to be 'Thu' . (Hint: Changing scores.columns[3] directly will not work, not to mention it would be a poor approach to hard code a location this way! Magic numbers be bad when referencing index or column locations. They may change!) Show the new scores data frame. This should be the same, except with Thurs changed to Thu.

In [11]:
scores = scores.rename(columns = {"Thurs": "Thu"})
scores

Mon   Tue    Wed    Thu   Fri Student
week_1  9.50  8.75   8.00  10.00  7.75     Bob
week_1  9.00  8.00  10.00   8.75  7.25    Jane
week_2  8.50  0.00   9.75   9.00  6.00     Bob
week_2  7.00  8.25   9.25   0.00  8.00    Jane
week_3  7.00  8.50   9.25   0.00  0.00     Bob
week_3  8.50  8.25   9.00   8.00  7.50    Jane
week_4  9.00  7.00   8.50   8.00  8.00     Bob
week_4  9.25  9.25   8.50   7.75  7.50    Jane

### 6) [P] Now, make a bigger change. Let's rename 'week_1' to be 'w1', 'week_2' to be 'w2', and so on.Also, suppose you decide to use only 2 letter abbreviations for the days. Rename the days in the column names to be 'Mo', 'Tu', 'We', 'Th', 'Fr'. Your new scores should appear as follows:

In [12]:
scores = scores.rename(columns = {"Mon":"Mo","Tue":"Tu","Wed":"We","Thu":"Th","Fri":"Fr"},index = {"week_1":"w1","week_2":"w2","week_3":"w3","week_4":"w4"})
scores

Mo    Tu     We     Th    Fr Student
w1  9.50  8.75   8.00  10.00  7.75     Bob
w1  9.00  8.00  10.00   8.75  7.25    Jane
w2  8.50  0.00   9.75   9.00  6.00     Bob
w2  7.00  8.25   9.25   0.00  8.00    Jane
w3  7.00  8.50   9.25   0.00  0.00     Bob
w3  8.50  8.25   9.00   8.00  7.50    Jane
w4  9.00  7.00   8.50   8.00  8.00     Bob
w4  9.25  9.25   8.50   7.75  7.50    Jane

### 7) [P] Compare the type of the expression scores['Mo'] vs. scores[['Mo']]. What is the difference?
(HINT: use the type() Python function on both.)
Read about the describe() method of data frames. (FYI - This is essentially the equivalent to the summary()
function in R.)

In [13]:
print(type(scores["Mo"]))
print(type(scores[["Mo"]]))

print("the first is a series and the second is a dataframe")

<class 'pandas.core.series.Series'>
<class 'pandas.core.frame.DataFrame'>
the first is a series and the second is a dataframe


### 8) [P] Demonstrate the describe() method on scores. What type does it return?
From your observations of
the output, which day had the largest standard deviation in quiz scores? (You can just hard code the answer to
this question. You'll write code to do this next.)

In [14]:
print(scores.describe())

print("\n this returns type: ",type(scores.describe()))

print("\nThursday has highest std dev")

             Mo      Tu         We         Th        Fr
count  8.000000  8.0000   8.000000   8.000000  8.000000
mean   8.468750  7.2500   9.031250   6.437500  6.500000
std    0.967669  3.0000   0.673974   4.037215  2.702512
min    7.000000  0.0000   8.000000   0.000000  0.000000
25%    8.125000  7.7500   8.500000   5.812500  6.937500
50%    8.750000  8.2500   9.125000   8.000000  7.500000
75%    9.062500  8.5625   9.375000   8.812500  7.812500
max    9.500000  9.2500  10.000000  10.000000  8.000000

 this returns type:  <class 'pandas.core.frame.DataFrame'>

Thursday has highest std dev


### 9) [P] While you could look at the output of describe() and infer information, it's more important to write code to output the specific results you are looking for. Write the Python code that shows the largest standard deviation and the day it occurred without using describe().
OK. Take a moment and look over that output. If you've been doing everything correctly, then that result should be
~4.0372 occurring on Thursday, and that is way too large! Think back to your stats course. What does standard
deviation mean? It's the deviation from the mean! That's a large spread in the data! Why? There are two missing
values represented by 0 that are being used to compute the deviation, and that's not what we want. Missing data
are usually not included with summary statistics (and many other calculations!) We need to deal with missing data
properly. So, instead, when we have missing values in data, we almost always use a na.nan value in its place to represent that the data is missing in that specific cell.

In [15]:
std = scores.std(axis = 0)
print("the largest standard deviation was about {0:.2f} on {1}".format(std.max(),std.idxmax()))


the largest standard deviation was about 4.04 on Th


### 10) [P] Write the code that changes all 0.0 entries to np.nan. 
(NOTE: You should be able to do this with just
one line of code using pandas selection techniques!) Then, show scores. All 0.0 entries should be replaced
with NaN in the output.
At this point, now you need to start paying attention to NaN values in your data. These represent missing values,
and are important to ignore in most calculations, especially preliminary EDA. We will learn about many techniques
for dealing with missing values soon. Many functions will have an optional parameter such as skipna to make
sure that whatever you are trying to do does not include these missing data. Pay attention to those parameters.
Most of the time, the default parameter values for handling NaN values is to skip or ignore them.

In [16]:
scores = scores.replace(0,np.nan) # could also have done inplace = True so we wouldn't have to write scores = 
scores

Mo    Tu     We     Th    Fr Student
w1  9.50  8.75   8.00  10.00  7.75     Bob
w1  9.00  8.00  10.00   8.75  7.25    Jane
w2  8.50   NaN   9.75   9.00  6.00     Bob
w2  7.00  8.25   9.25    NaN  8.00    Jane
w3  7.00  8.50   9.25    NaN   NaN     Bob
w3  8.50  8.25   9.00   8.00  7.50    Jane
w4  9.00  7.00   8.50   8.00  8.00     Bob
w4  9.25  9.25   8.50   7.75  7.50    Jane

### 11) [P] Show the output of describe() again. 
Yes, this is better! Note how much more informative this output
is when properly dealing with missing data. Copy your code from above to output the day with the largest
standard deviation in quiz scores. You should have a substantially different result than you did before!

In [17]:
print(scores.describe())
std = scores.std(axis = 0)
print("\nthe largest standard deviation was about {0:.2f} on {1}".format(std.max(),std.idxmax()))

             Mo        Tu         We         Th        Fr
count  8.000000  7.000000   8.000000   6.000000  7.000000
mean   8.468750  8.285714   9.031250   8.583333  7.428571
std    0.967669  0.698638   0.673974   0.846562  0.687906
min    7.000000  7.000000   8.000000   7.750000  6.000000
25%    8.125000  8.125000   8.500000   8.000000  7.375000
50%    8.750000  8.250000   9.125000   8.375000  7.500000
75%    9.062500  8.625000   9.375000   8.937500  7.875000
max    9.500000  9.250000  10.000000  10.000000  8.000000

the largest standard deviation was about 0.97 on Mo


### 12) [P] Show the output of scores.values. What does the .values attribute do?

In [18]:
print(scores.values)
print("\n this outputs scores as a 2d numpy array")

[[9.5 8.75 8.0 10.0 7.75 'Bob']
 [9.0 8.0 10.0 8.75 7.25 'Jane']
 [8.5 nan 9.75 9.0 6.0 'Bob']
 [7.0 8.25 9.25 nan 8.0 'Jane']
 [7.0 8.5 9.25 nan nan 'Bob']
 [8.5 8.25 9.0 8.0 7.5 'Jane']
 [9.0 7.0 8.5 8.0 8.0 'Bob']
 [9.25 9.25 8.5 7.75 7.5 'Jane']]

 this outputs scores as a 2d numpy array


### 13) [P] What is the mean quiz score for each day? Do not use describe()

In [19]:
scores.mean() # could put in axis = 0  but it will default to that because it is finding means over the index/rows to get a mean for each column
# also skips nan by default

Mo    8.468750
Tu    8.285714
We    9.031250
Th    8.583333
Fr    7.428571
dtype: float64

### 14) [P] What is the mean quiz score over the entire dataset?
(HINT: there are many ways to do this, and none
should involve writing any loops! The only right answer is 8.375. If you aren't getting that, then you are not
handling NaN values correctly. You only want the mean over all scores that are NOT NaN!)

In [20]:
# get the mean of all the means 
scores.loc[:,"Mo":"Fr"].stack().mean() 

#in order to not overweight columns with more nan values (less actual values), we cant just do .mean().mean(). instead, we use loc to not include names
# then use stack() to make it 1 dimensional. a series. then we can get the mean with every nominal value having equal weight. skipna is True by default so we don't need to include that
# because we are converting it to a series, we can not use the numeric_only arg for .mean(), so we have to do the .loc

8.375

### 15) [P] Now, show the mean for each week over all students. 
Again, you are not counting missing quizzes into the
average. (Yes, technically you could write a loop to compute this exhaustively. However, you should rarely
need to do that. Use pandas!)
You have the correct result if you report the following (shown with 5 places of significant digits):

In [21]:
print("Mean for each week:")
scores.mean(axis = 1).groupby(level = 0).mean()
# we get the mean for each week, separated, then combine means for same week by grouping by index and combining by taking the mean

Mean for each week:


w1    8.70000
w2    8.21875
w3    8.25000
w4    8.27500
dtype: float64

### 16) [P] Select the data frame for only Monday and Student variables using .loc, and again using .iloc

In [22]:
print("select monday and student using loc : \n" ,scores.loc[:,["Mo","Student"]])
print("select monday and student using iloc : \n" ,scores.iloc[:,[0,-1]])

select monday and student using loc : 
       Mo Student
w1  9.50     Bob
w1  9.00    Jane
w2  8.50     Bob
w2  7.00    Jane
w3  7.00     Bob
w3  8.50    Jane
w4  9.00     Bob
w4  9.25    Jane
select monday and student using iloc : 
       Mo Student
w1  9.50     Bob
w1  9.00    Jane
w2  8.50     Bob
w2  7.00    Jane
w3  7.00     Bob
w3  8.50    Jane
w4  9.00     Bob
w4  9.25    Jane


### 17) [P] Show a new DataFrame containing the week as an index (like the original), but the mean of their scoresand the count of the quizzes they took. Your result should look as follows:

In [23]:
students = scores.loc[:,"Student"]
count = scores.count(axis = 1).rename("count")
means = scores.mean(axis = 1 ).rename("mean")
                            
pd.concat([students,count,means],axis = 1)

# axis  = 1 means we are going through columns
# axis =1 for mean gives a mean for a row because it goes through all of that row's columns
# we make each part into a series then concatenate by column (we add each as a new column)

Student  count    mean
w1     Bob      6  8.8000
w1    Jane      6  8.6000
w2     Bob      5  8.3125
w2    Jane      5  8.1250
w3     Bob      4  8.2500
w3    Jane      6  8.2500
w4     Bob      6  8.1000
w4    Jane      6  8.4500

### 18) [P] Show the number of times Friday's score was greater than 7.25 for each student. Start by selecting your data with the [ ] operator. Your result should look as follows:

In [68]:
#worked with Matt McLaughlin
friday = scores.loc[:,["Fr","Student"]] #get only the things we are using (Friday scores and student names)
friday.loc[:,"Fr"]  = friday.loc[:,"Fr"] > 7.25 # make scores booleans
friday.groupby(by = "Student").sum() # sum trues for each name by grouping booleans by student then summing

Fr
Student     
Bob      2.0
Jane     3.0

### 19) [P] Use the where() method on scores to repeat the previous exercise. 
(HINT: You might need to use
dropna()). You should have the same output

In [87]:
#worked with Matt McLaughlin
friday = scores.loc[:,["Fr","Student"]] #get only the things we are using (Friday scores and student names)
friday.where(friday["Fr"] > 7.25).groupby(by = "Student").count()

#dropna not necessary in this specific case, but would be good to include before .groupby for more general use

Fr
Student    
Bob       2
Jane      3

20) [P] Select the scores that were greater than the mean score over the entire dataset (without the missing
values). Again, no hard coding values here (i.e. do NOT hard code the mean, but recompute it. Then, report
the number of scores for each day that exceeded that global mean. Your result should look like:

This suggests that Wednesday seems to have the best quiz performance, whereas Friday... not so much.

In [ ]:
#worked with Matt McLaughlin
scoresNoName = scores.loc[:,"Mo":"Fr"]
totalMean = scoresNoName.stack().mean()
scoresNoName.where(scoresNoName > totalMean).count()